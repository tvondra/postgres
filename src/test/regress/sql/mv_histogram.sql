-- data type passed by value
CREATE TABLE mv_histogram (
    a INT,
    b INT,
    c INT
);

-- unknown column
CREATE STATISTICS s1 ON mv_histogram (unknown_column) WITH (histogram);

-- single column
CREATE STATISTICS s1 ON mv_histogram (a) WITH (histogram);

-- single column, duplicated
CREATE STATISTICS s1 ON mv_histogram (a, a) WITH (histogram);

-- two columns, one duplicated
CREATE STATISTICS s1 ON mv_histogram (a, a, b) WITH (histogram);

-- unknown option
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (unknown_option);

-- missing histogram statistics
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (dependencies, max_buckets 200);

-- invalid max_buckets value / too low
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (mcv, max_buckets 10);

-- invalid max_buckets value / too high
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (mcv, max_buckets 100000);

-- correct command
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (histogram);

-- random data (no functional dependencies)
INSERT INTO mv_histogram
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);

ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

TRUNCATE mv_histogram;

-- a => b, a => c, b => c
INSERT INTO mv_histogram
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);

ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

TRUNCATE mv_histogram;

-- a => b, a => c
INSERT INTO mv_histogram
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

TRUNCATE mv_histogram;

-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mv_histogram
     SELECT i/10000, i/20000, i/40000 FROM generate_series(1,1000000) s(i);
CREATE INDEX hist_idx ON mv_histogram (a, b);
ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM mv_histogram WHERE a = 10 AND b = 5;

DROP TABLE mv_histogram;

-- varlena type (text)
CREATE TABLE mv_histogram (
    a TEXT,
    b TEXT,
    c TEXT
);

CREATE STATISTICS s2 ON mv_histogram (a, b, c) WITH (histogram);

-- random data (no functional dependencies)
INSERT INTO mv_histogram
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);

ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

TRUNCATE mv_histogram;

-- a => b, a => c, b => c
INSERT INTO mv_histogram
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);

ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

TRUNCATE mv_histogram;

-- a => b, a => c
INSERT INTO mv_histogram
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

TRUNCATE mv_histogram;

-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mv_histogram
     SELECT i/10000, i/20000, i/40000 FROM generate_series(1,1000000) s(i);
CREATE INDEX hist_idx ON mv_histogram (a, b);
ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM mv_histogram WHERE a = '10' AND b = '5';

TRUNCATE mv_histogram;

-- check explain (expect bitmap index scan, not plain index scan) with NULLs
INSERT INTO mv_histogram
     SELECT
       (CASE WHEN i/10000 = 0 THEN NULL ELSE i/10000 END),
       (CASE WHEN i/20000 = 0 THEN NULL ELSE i/20000 END),
       (CASE WHEN i/40000 = 0 THEN NULL ELSE i/40000 END)
     FROM generate_series(1,1000000) s(i);
ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

EXPLAIN (COSTS off)
 SELECT * FROM mv_histogram WHERE a IS NULL AND b IS NULL;

DROP TABLE mv_histogram;

-- NULL values (mix of int and text columns)
CREATE TABLE mv_histogram (
    a INT,
    b TEXT,
    c INT,
    d TEXT
);

CREATE STATISTICS s3 ON mv_histogram (a, b, c, d) WITH (histogram);

INSERT INTO mv_histogram
     SELECT
         mod(i, 100),
         (CASE WHEN mod(i, 200) = 0 THEN NULL ELSE mod(i,200) END),
         mod(i, 400),
         (CASE WHEN mod(i, 300) = 0 THEN NULL ELSE mod(i,600) END)
     FROM generate_series(1,10000) s(i);

ANALYZE mv_histogram;

SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;

DROP TABLE mv_histogram;
