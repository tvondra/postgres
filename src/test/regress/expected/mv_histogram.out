-- data type passed by value
CREATE TABLE mv_histogram (
    a INT,
    b INT,
    c INT
);
-- unknown column
CREATE STATISTICS s1 ON mv_histogram (unknown_column) WITH (histogram);
ERROR:  column "unknown_column" referenced in statistics does not exist
-- single column
CREATE STATISTICS s1 ON mv_histogram (a) WITH (histogram);
ERROR:  multivariate stats require 2 or more columns
-- single column, duplicated
CREATE STATISTICS s1 ON mv_histogram (a, a) WITH (histogram);
ERROR:  duplicate column name in statistics definition
-- two columns, one duplicated
CREATE STATISTICS s1 ON mv_histogram (a, a, b) WITH (histogram);
ERROR:  duplicate column name in statistics definition
-- unknown option
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (unknown_option);
ERROR:  unrecognized STATISTICS option "unknown_option"
-- missing histogram statistics
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (dependencies, max_buckets 200);
ERROR:  option 'histogram' is required by other options(s)
-- invalid max_buckets value / too low
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (mcv, max_buckets 10);
ERROR:  minimum number of buckets is 128
-- invalid max_buckets value / too high
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (mcv, max_buckets 100000);
ERROR:  maximum number of buckets is 16384
-- correct command
CREATE STATISTICS s1 ON mv_histogram (a, b, c) WITH (histogram);
-- random data (no functional dependencies)
INSERT INTO mv_histogram
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

TRUNCATE mv_histogram;
-- a => b, a => c, b => c
INSERT INTO mv_histogram
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

TRUNCATE mv_histogram;
-- a => b, a => c
INSERT INTO mv_histogram
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

TRUNCATE mv_histogram;
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mv_histogram
     SELECT i/10000, i/20000, i/40000 FROM generate_series(1,1000000) s(i);
CREATE INDEX hist_idx ON mv_histogram (a, b);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM mv_histogram WHERE a = 10 AND b = 5;
                 QUERY PLAN                 
--------------------------------------------
 Bitmap Heap Scan on mv_histogram
   Recheck Cond: ((a = 10) AND (b = 5))
   ->  Bitmap Index Scan on hist_idx
         Index Cond: ((a = 10) AND (b = 5))
(4 rows)

DROP TABLE mv_histogram;
-- varlena type (text)
CREATE TABLE mv_histogram (
    a TEXT,
    b TEXT,
    c TEXT
);
CREATE STATISTICS s2 ON mv_histogram (a, b, c) WITH (histogram);
-- random data (no functional dependencies)
INSERT INTO mv_histogram
     SELECT mod(i, 111), mod(i, 123), mod(i, 23) FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

TRUNCATE mv_histogram;
-- a => b, a => c, b => c
INSERT INTO mv_histogram
     SELECT i/10, i/100, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

TRUNCATE mv_histogram;
-- a => b, a => c
INSERT INTO mv_histogram
     SELECT i/10, i/150, i/200 FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

TRUNCATE mv_histogram;
-- check explain (expect bitmap index scan, not plain index scan)
INSERT INTO mv_histogram
     SELECT i/10000, i/20000, i/40000 FROM generate_series(1,1000000) s(i);
CREATE INDEX hist_idx ON mv_histogram (a, b);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM mv_histogram WHERE a = '10' AND b = '5';
                         QUERY PLAN                         
------------------------------------------------------------
 Bitmap Heap Scan on mv_histogram
   Recheck Cond: ((a = '10'::text) AND (b = '5'::text))
   ->  Bitmap Index Scan on hist_idx
         Index Cond: ((a = '10'::text) AND (b = '5'::text))
(4 rows)

TRUNCATE mv_histogram;
-- check explain (expect bitmap index scan, not plain index scan) with NULLs
INSERT INTO mv_histogram
     SELECT
       (CASE WHEN i/10000 = 0 THEN NULL ELSE i/10000 END),
       (CASE WHEN i/20000 = 0 THEN NULL ELSE i/20000 END),
       (CASE WHEN i/40000 = 0 THEN NULL ELSE i/40000 END)
     FROM generate_series(1,1000000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

EXPLAIN (COSTS off)
 SELECT * FROM mv_histogram WHERE a IS NULL AND b IS NULL;
                    QUERY PLAN                     
---------------------------------------------------
 Bitmap Heap Scan on mv_histogram
   Recheck Cond: ((a IS NULL) AND (b IS NULL))
   ->  Bitmap Index Scan on hist_idx
         Index Cond: ((a IS NULL) AND (b IS NULL))
(4 rows)

DROP TABLE mv_histogram;
-- NULL values (mix of int and text columns)
CREATE TABLE mv_histogram (
    a INT,
    b TEXT,
    c INT,
    d TEXT
);
CREATE STATISTICS s3 ON mv_histogram (a, b, c, d) WITH (histogram);
INSERT INTO mv_histogram
     SELECT
         mod(i, 100),
         (CASE WHEN mod(i, 200) = 0 THEN NULL ELSE mod(i,200) END),
         mod(i, 400),
         (CASE WHEN mod(i, 300) = 0 THEN NULL ELSE mod(i,600) END)
     FROM generate_series(1,10000) s(i);
ANALYZE mv_histogram;
SELECT hist_enabled, hist_built
  FROM pg_mv_statistic WHERE starelid = 'mv_histogram'::regclass;
 hist_enabled | hist_built 
--------------+------------
 t            | t
(1 row)

DROP TABLE mv_histogram;
